### PR TITLE
Legend Authentication Refactoring - Part 1

### PR DESCRIPTION
**Pull request by epsstan**
#### What type of PR is this?

Enhancement

#### What does this PR do / why is it needed ?

Introduces new APIs and modules that will enable authentication constructs/code to be re-used across the platform.

#### Which issue(s) this PR fixes:

N/A

#### Other notes for reviewers:

Refactoring of authentication is a big code change that will eventually touch many modules. 
To reduce the size of the PR, this refactoring will be broken down and  merged as many small PRs.

#### Does this PR introduce a user-facing change?
N/A

### CODE REVIEW

File: legend-engine-authentication-grammar/src/main/antlr4/org/finos/legend/engine/language/pure/grammar/from/antlr4/authentication/AuthenticationLexerGrammar.g4 [hunk 1]
```diff
@@ -0,0 +1,36 @@
+lexer grammar AuthenticationLexerGrammar;
+
+import CoreLexerGrammar;
+
+// ---------------- OAuth Authentication
+
+OAUTH_AUTHENTICATION:                                       'OAuth';
+GRANT_TYPE:                                                 'grantType';
+CLIENT_ID:                                                  'clientId';
+CLIENT_SECRET_VAULT_REFERENCE:                              'clientSecretVaultReference';
```
Comment (jinanisha) [inline legend-engine-authentication-grammar/src/main/antlr4/org/finos/legend/engine/language/pure/grammar/from/antlr4/authentication/AuthenticationLexerGrammar.g4:10]: clientSecret to take a Vault Credential and not directly a vault reference 
clientSecret : CredentialVault { ref = "ref1"}

**Comment by TheJuanAndOnly99** [comment]
@cla-bot[bot] check

**Comment by TheJuanAndOnly99** [comment]
Hi @epsstan you're all set.

File: legend-engine-xt-authentication-protocol/src/main/java/org/finos/legend/engine/protocol/pure/v1/model/packageableElement/authentication/vault/aws/StaticAWSCredentials.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
+package org.finos.legend.engine.protocol.pure.v1.model.packageableElement.authentication.vault.aws;
```
Comment (rafaelbey) [inline legend-engine-xt-authentication-protocol/src/main/java/org/finos/legend/engine/protocol/pure/v1/model/packageableElement/authentication/vault/aws/StaticAWSCredentials.java:1]: Please add the copyright header

File: legend-engine-xt-authentication-example/pom.xml [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+        <version>3.34.1-SNAPSHOT</version>
+    </parent>
+    <modelVersion>4.0.0</modelVersion>
+
+    <artifactId>legend-engine-xt-authentication-example</artifactId>
+    <packaging>jar</packaging>
+    <name>Legend Engine - XT - Authentication - Example</name>
+
+    <build>
+        <plugins>
```
Comment (rafaelbey) [inline legend-engine-xt-authentication-example/pom.xml:31]: Do we need all these plugins?  I think some of these are defaulted by the parent pom already.  And for example, test-jar add another artifact, so we should make sure we have test code to share.

**Review by gs-ssh16 — APPROVED** [review]

**Pull request merged**